### PR TITLE
PHP 8: clean up `Http` package

### PR DESCRIPTION
**Pull request by distantnative**

### CODE REVIEW

File: src/Http/Identifiers.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * A wrapper for URL query and params
+ * @internal
+ *
+ * @package   Kirby Http
+ * @author    Nico Hoffmann <nico@getkirby.com>
+ * @link      https://getkirby.com
+ * @copyright Bastian Allgeier
+ * @license   https://opensource.org/licenses/MIT
+ */
+abstract class Identifiers extends Obj
```
Comment (distantnative) [inline src/Http/Identifiers.php]: @bastianallgeier @lukasbestle if you have any idea for a better name, please let me know

Comment (lukasbestle) [reply src/Http/Identifiers.php]: Maybe `KeyValueData`?

Comment (distantnative) [reply src/Http/Identifiers.php]: Mmm that sounds super general. Maybe it's not even worth it to have a common parent class for keeping it DRY just a little bit more. Maybe you creating both classes separately with these few methods existing in both would be the cleaner way after all.

Comment (afbora) [reply src/Http/Identifiers.php]: I am not sure. I'm just thinking out loud. Don't you think it is more appropriate to use `Trait` instead `abstract class`  in this context? 🤔

Comment (distantnative) [reply src/Http/Identifiers.php]: @afbora what would be the benefit?

Comment (afbora) [reply src/Http/Identifiers.php]: Great question 😄 To be honest, I can't show any obvious benefit. It just made more sense to me to use trait in terms of its function.

Does having several methods be similar require inheritance? Even the constructor of `Query` and `Params` is different anyway. `Identifiers::__construct` didn't seem very functional. As a result `Identifiers` class seemed like an unnecessary layer of inheritance to me. Not important. It's just that this extra layer was a bit more tiring when reviewing the code.

Comment (distantnative) [reply src/Http/Identifiers.php]: Yea but that's be the same as a trait. Really leaning much more towards putting the same code just in the two classes. Tiny bit less DRY but so much better to read I think.

Comment (lukasbestle) [reply src/Http/Identifiers.php]: I agree.

Comment (distantnative) [reply src/Http/Identifiers.php]: ✅

File: src/Http/Cookie.php [hunk 2]
```diff
@@ -122,7 +120,7 @@ public static function forever(string $key, string $value, array $options = []):
 	 *                             if the cookie has not been found
 	 * @return mixed The found value
```
Comment (afbora) [inline src/Http/Cookie.php]:
```suggestion
	 * @return string|array|null The found value
```

File: src/Http/Cookie.php [hunk 3]
```diff
@@ -122,7 +120,7 @@ public static function forever(string $key, string $value, array $options = []):
...
 	 */
-	public static function get(string $key = null, string $default = null)
+	public static function get(string|null $key = null, string|null $default = null): mixed
```
Comment (afbora) [inline src/Http/Cookie.php]:
```suggestion
	public static function get(string|null $key = null, string|null $default = null): string|array|null
```

File: src/Http/Url.php [hunk 4]
```diff
@@ -219,45 +172,32 @@ public static function short($url = null, int $length = 0, bool $base = false, s
...
 
 	/**
 	 * Smart resolver for internal and external urls
-	 *
-	 * @param string $path
-	 * @param mixed $options
-	 * @return string
 	 */
-	public static function to(string $path = null, $options = null): string
+	public static function to(string|null $path = null, array|null $options = null): string
```
Comment (afbora) [inline src/Http/Url.php]: `Uri::__construct` second parameter (for `$options` argument) not nullable.

File: src/Http/Visitor.php [hunk 5]
```diff
@@ -65,11 +44,8 @@ public function __construct(array $arguments = [])
 	 * Sets the accepted language if
 	 * provided or returns the user's
 	 * accepted language otherwise
-	 *
-	 * @param string|null $acceptedLanguage
-	 * @return \Kirby\Toolkit\Obj|\Kirby\Http\Visitor|null
 	 */
-	public function acceptedLanguage(string $acceptedLanguage = null)
+	public function acceptedLanguage(string|null $acceptedLanguage = null): Obj|Visitor|null
```
Comment (afbora) [inline src/Http/Visitor.php]:
```suggestion
	public function acceptedLanguage(string|null $acceptedLanguage = null): static|Obj|null
```

**Review by afbora — APPROVED** [review]

File: src/Http/Environment.php [hunk 6]
```diff
@@ -233,11 +194,8 @@ public function detect(array $options = null, array $info = null): array
 	/**
 	 * Sets the host name, port, path and protocol from the
 	 * fixed list of allowed URLs
-	 *
-	 * @param array|string $allowed
-	 * @return void
 	 */
-	protected function detectAllowed($allowed): void
+	protected function detectAllowed(array|string|object $allowed): void
```
Comment (lukasbestle) [inline src/Http/Environment.php:198]: Why `object`? Won't this be wrapped into `[object]` and then tried to be passed to `new Uri()` (which only accepts `array|string`)?

File: src/Http/Response.php [hunk 7]
```diff
@@ -206,13 +175,9 @@ public static function file(string $file, array $props = [])
...
 	 * @since 3.7.0
 	 *
-	 * @param string $url
-	 * @param int $code
-	 * @return void
-	 *
 	 * @codeCoverageIgnore
 	 */
-	public static function go(string $url = '/', int $code = 302)
+	public static function go(string $url = '/', int $code = 302): void
```
Comment (lukasbestle) [inline src/Http/Response.php:180]: We could add a `TODO` comment to change the return type to `never` once we drop support for PHP 8.0.

File: src/Http/Router.php [hunk 8]
```diff
@@ -166,13 +149,8 @@ public static function execute(string|null $path = null, string $method = 'GET',
...
 	 * find matches and return all the found
 	 * arguments in the path.
-	 *
-	 * @param string $path
-	 * @param string $method
-	 * @param array $ignore
-	 * @return \Kirby\Http\Route|null
 	 */
-	public function find(string $path, string $method, array $ignore = null)
+	public function find(string $path, string $method, array|null $ignore = null): Route|null
```
Comment (lukasbestle) [inline src/Http/Router.php:153]: How can the return type be `null`?

File: src/Http/Router.php [hunk 9]
```diff
@@ -120,23 +109,23 @@ public function call(string $path = null, string $method = 'GET', Closure $callb
...
+			if (is_a($this->beforeEach, Closure::class) === true) {
 				($this->beforeEach)($route, $path, $method);
 			}
 
 			try {
 				if ($callback) {
 					$result = $callback($route);
 				} else {
-					$result = $route->action()->call($route, ...$route->arguments());
+					$result = $route?->action()->call($route, ...$route->arguments());
```
Comment (lukasbestle) [inline src/Http/Router.php:120]: Followup to the previous comment: I think it can never be `null`.

**Pull request merged**